src/ripple/shamap/impl/SHAMapInnerNode.cpp

```diff
@@ -33,26 +33,718 @@
+}
+
+[[nodiscard]] inline std::uint8_t
+SHAMapInnerNode::TaggedPointer::numAllocated() const
+{
+    return boundaries[tp_ & tagMask];
+}
+
+[[nodiscard]] inline bool
+SHAMapInnerNode::TaggedPointer::isDense() const
+{
+    return (tp_ & tagMask) == boundaries.size() - 1;
+}
+
+[[nodiscard]] inline std::
+    tuple<std::uint8_t, SHAMapHash*, std::shared_ptr<SHAMapAbstractNode>*>
+    SHAMapInnerNode::TaggedPointer::getHashesAndChildren() const
+{
+    auto const [tag, ptr] = decode();
+    auto const hashes = reinterpret_cast<SHAMapHash*>(ptr);
+    std::uint8_t numAllocated = boundaries[tag];
+    auto const children =
+        reinterpret_cast<std::shared_ptr<SHAMapAbstractNode>*>(
+            hashes + numAllocated);
+    return {numAllocated, hashes, children};
+};
+
+[[nodiscard]] inline SHAMapHash*
+SHAMapInnerNode::TaggedPointer::getHashes() const
+{
+    return reinterpret_cast<SHAMapHash*>(tp_ & ptrMask);
+};
+
+[[nodiscard]] inline std::shared_ptr<SHAMapAbstractNode>*
+SHAMapInnerNode::TaggedPointer::getChildren() const
+{
+    auto [unused1, unused2, result] = getHashesAndChildren();
+    return result;
+};
+
 std::mutex SHAMapInnerNode::childLock;
 
+SHAMapInnerNode::SHAMapInnerNode(
+    std::uint32_t cowid,
+    std::uint8_t numAllocatedChildren)
+    : SHAMapAbstractNode(cowid), hashesAndChildren_(numAllocatedChildren)
+{
+}
+
+template <class F>
+void
+SHAMapInnerNode::iterChildren(F&& f) const
+{
+    hashesAndChildren_.iterChildren(isBranch_, std::forward<F>(f));
+}
+
+template <class F>
+void
+SHAMapInnerNode::iterNonEmptyChildIndexes(F&& f) const
+{
+    hashesAndChildren_.iterNonEmptyChildIndexes(isBranch_, std::forward<F>(f));
+}
+
+void
+SHAMapInnerNode::resizeChildArrays(std::uint8_t toAllocate)
+{
+    hashesAndChildren_ =
+        TaggedPointer(std::move(hashesAndChildren_), isBranch_, toAllocate);
+}
+
+std::optional<int>
+SHAMapInnerNode::getChildIndex(int i) const
+{
+    return hashesAndChildren_.getChildIndex(isBranch_, i);
+}
+
 std::shared_ptr<SHAMapAbstractNode>
 SHAMapInnerNode::clone(std::uint32_t cowid) const
 {
-    auto p = std::make_shared<SHAMapInnerNode>(cowid);
+    auto const branchCount = getBranchCount();
+    auto const thisIsSparse = !hashesAndChildren_.isDense();
+    auto p = std::make_shared<SHAMapInnerNode>(cowid, branchCount);
     p->hash_ = hash_;
-    p->mIsBranch = mIsBranch;
-    p->mFullBelowGen = mFullBelowGen;
-    p->mHashes = mHashes;
+    p->isBranch_ = isBranch_;
+    p->fullBelowGen_ = fullBelowGen_;
+    SHAMapHash *cloneHashes, *thisHashes;
+    std::shared_ptr<SHAMapAbstractNode>*cloneChildren, *thisChildren;
+    // structured bindings can't be captured in c++ 17; use tie instead
+    std::tie(std::ignore, cloneHashes, cloneChildren) =
+        p->hashesAndChildren_.getHashesAndChildren();
+    std::tie(std::ignore, thisHashes, thisChildren) =
+        hashesAndChildren_.getHashesAndChildren();
+
+    if (thisIsSparse)
+    {
+        // sparse
```

Comment:
Comment here and below seems unnecessary.

Reply:
Fixed in `2fa4a8d59e [fold] Doxygen comments and misc cleanups 1:`